src/proto/h1/date.rs

```diff
@@ -52,18 +52,18 @@ impl CachedDate {
     }
 
     fn check(&mut self) {
-        let now = time::get_time();
+        let now = OffsetDateTime::now();
         if now > self.next_update {
             self.update(now);
         }
     }
 
-    fn update(&mut self, now: time::Timespec) {
+    fn update(&mut self, now: OffsetDateTime) {
         self.pos = 0;
-        let _ = write!(self, "{}", time::at_utc(now).rfc822());
+        let _ = write!(self, "{}", now.format("%a, %d %b %Y %H:%M:%S GMT"));
         debug_assert!(self.pos == DATE_VALUE_LENGTH);
-        self.next_update = now + Duration::seconds(1);
-        self.next_update.nsec = 0;
+        let nanosecond = self.next_update.time().nanosecond();
+        self.next_update = now + Duration::seconds(1) - Duration::nanoseconds(nanosecond as i64);
```

Comment:
Easier to just add the number of nanoseconds directly. It may be worth checking the optimized output of this, though, as the compiler may be able to optimize away the `x - x`.

```suggestion
        self.next_update = now + Duration::nanoseconds(1_000_000_000 - now.nanosecond() as i64);
```